### PR TITLE
[Substrait to Velox] agg support and add q6 test #1554

### PR DESCRIPTION
**Pull request by zhejiangxiaomai**
- Supported TPC-H Q1 related conversions from Substrait to Velox, including:
    avg, count, row construct, plus, minus
- Supported using a pre-built plan node as the data source node.
- Update the JSON file used by UT, now UT can 100% passed. result is
 [==========] 10 tests from 3 test suites ran. (111 ms total)
 [  PASSED  ] 10 tests.

### CODE REVIEW

**Comment by mbasmanova** [comment]
@zhejiangxiaomai This PR is very large, more than 1K lines of changes. Would you update the PR description to explain the changes? Is there a way to break up this PR into a few smaller ones? 

I see changes to .proto files. Are these accidental?

File: velox/substrait/SubstraitToVeloxPlan.h [hunk 1]
```diff
@@ -109,18 +109,54 @@ class SubstraitVeloxPlanConverter {
   /// name>:<arg_type0>_<arg_type1>_..._<arg_typeN>
   const std::string& findFunction(uint64_t id) const;
 
+  /// Used to insert certain plan node as input. The plan node
+  /// id will start from the setted one.
+  void insertInputNode(
+      uint64_t inputIdx,
+      const std::shared_ptr<const core::PlanNode>& inputNode,
```
Comment (mbasmanova) [inline velox/substrait/SubstraitToVeloxPlan.h:116]: use PlanNodePtr for readability

**Review by mbasmanova — COMMENTED** [review]
@zhejiangxiaomai The logic to flatten partial aggregation results and combine them back is very Spark-specific and therefore doesn't belong to a generic Velox-to-Substrait converter. Let's figure out how to put this logic somewhere else.

**Comment by zhejiangxiaomai** [comment]
> @mbasmanova 
thanks for review my PR. In fact, JSON changes are necessary. because [#1212](https://github.com/facebookincubator/velox/pull/1212) add "cc_enable_arenas" option. So we need update JSON to pass all UT, otherwise, we can not get correct plan. In addition to those JSON changes, the real code modification is less than 500 lines. Mainly involved SubstraitToVeloxPlan.cpp  and SubstraitToVeloxPlan.h [#1554](https://github.com/facebookincubator/velox/pull/1554)

File: velox/substrait/SubstraitToVeloxPlan.h (hunk 1 above)
Comment (zhejiangxiaomai) [reply velox/substrait/SubstraitToVeloxPlan.h]: fixed

**Comment by zhejiangxiaomai** [comment]
> @zhejiangxiaomai The logic to flatten partial aggregation results and combine them back is very Spark-specific and therefore doesn't belong to a generic Velox-to-Substrait converter. Let's figure out how to put this logic somewhere else.

can you give some suggestions？

**Comment by mbasmanova** [comment]
@zhejiangxiaomai Thank you for updating the PR description. I see it contains 3 changes that appear independent. Would it be possible to submit each change as a separate PR?

**Comment by mbasmanova** [comment]
The PR description is still lacking the explanation of the logic used to flatten and combine intermediate aggregate results. I can only guess what that logic is for and how it works.

I assume the problem is the mismatch between intermediate types for the aggregate functions in Velox and Spark. 

For example, avg function in Velox has intermediate type ROW(BIGINT, DOUBLE - a struct with two fields: count and sum. in Spark intermediate type for the avg function is represented as two separate columns: count and sum. Given SELECT a, avg(b) FROM t query with NTEGER columns a and b, Velox plan looks like:

  - PartialAggregation -> a: INTEGER, avg: ROW(BIGINT, DOUBLE) 

while Spark plan looks like

  - PartialAggregation -> a: INTEGER, avg_count:BIGINT, avg_count:DOUBLE  


In general, in Velox all aggregate functions use single-column intermediate type. For some functions this type is simple, e.g. BIGINT for sum(BIGINT). For others, it is a struct with two or more fields.

One option to bridge this gap is to adjust the Spark plan before converting to Substrait and adjust again after conversion from Substrait.

BTW, it would be useful to check whether Substrait has any assumptions about the output type of the PartialAggregation nodes. In Velox, this node always contains grouping keys followed by interemediate result columns, one column per aggregate function. 

CC: @beroyfb @miaoever

**Comment by mbasmanova** [comment]
My reading of Substrait specification suggests that aggregate functions there are expected to use a single column intermediate and final result.

https://substrait.io/expressions/aggregate_functions/ 

-- algebra.proto
```
message AggregateFunction {
  // points to a function_anchor defined in this plan
  uint32 function_reference = 1;
  repeated Expression args = 2;
  repeated SortField sorts = 3;
  AggregationPhase phase = 4;
  Type output_type = 5;
}

```

-- extensions/functions_arithmetic.yaml
```
  - name: "avg"
    description: Average a set of values. For integral types, this truncates partial values.
    impls:
      - args:
          - options: [ SILENT, SATURATE, ERROR ]
            required: false
          - value: i8
        nullability: DECLARED_OUTPUT
        decomposable: MANY
        intermediate: "STRUCT<i64,i64>"
        return: i8?
```

**Comment by zhejiangxiaomai** [comment]
> @zhejiangxiaomai Thank you for updating the PR description. I see it contains 3 changes that appear independent. Would it be possible to submit each change as a separate PR?

OK, I  will split into two PRs

**Comment by rui-mo** [comment]
Hi @mbasmanova, thanks for your comment on the aggregation. As you mentioned, the intermediate type for Velox and Spark is different. For now, we are checking whether a projection to combine intermediate columns into one row-type column is needed. If yes, the intermediate columns will be combined before aggregation. If handing them during plan generation, an extra projection can be added as the child of aggregation in Substrait plan to specify the column combination (by using row construct function). How do you think so?

**Comment by mbasmanova** [comment]
@rui-mo Hi Rui. I do not quite understand your proposal above. It is not desirable to add logic to plan converter to alter the shape of the plan, i.e. we'd want the number and types of nodes in Velox and Substrait plans to match.

**Comment by rui-mo** [comment]
> @rui-mo Hi Rui. I do not quite understand your proposal above. It is not desirable to add logic to plan converter to alter the shape of the plan, i.e. we'd want the number and types of nodes in Velox and Substrait plans to match.

Yes, we are not proposing to alter the plan in the plan converter, but to generate a new plan for the use of plan converter. Becuase spark have 'sum' and 'count' outputs for partial avg, we need to let Velox help us to combine them for the computing of final avg. Therefore, in Gluten's plan generation, we can add a project rel before aggregate rel to represent this. In this way, project rel -> project node, aggregation rel -> aggregation node, and the numbers of Substrait rel and Velox node are matched.

**Comment by mbasmanova** [comment]
@rui-mo Rui, thank you for clarifying. This sounds good to me. Looking forward to an updated PR.

**Comment by zhejiangxiaomai** [comment]
Hi, @mbasmanova  we make a new PR named [#1978](https://github.com/facebookincubator/velox/pull/1978) those conversation.

**Pull request closed without merging**